From 4ff93e2e2d5253ac8aa773861e15dd41c0b5d4fd Mon Sep 17 00:00:00 2001
From: Fedora X Ninjas <x@fedoraproject.org>
Date: Wed, 12 Dec 2012 15:15:38 +1000
Subject: [PATCH] autoconfig: fixup tell changed so randr clients can tell

This lets the gnome applet update if a usb device appears/disappears
---
 hw/xfree86/common/xf86platformBus.c |  7 +++++--
 hw/xfree86/modes/xf86Crtc.c         |  2 ++
 hw/xfree86/modes/xf86RandR12.c      |  2 ++
 randr/randr.c                       | 24 +++++++++++++++++++++++-
 randr/randrstr.h                    |  4 ++++
 randr/rrcrtc.c                      |  2 +-
 randr/rrinfo.c                      |  2 +-
 randr/rroutput.c                    |  2 +-
 randr/rrscreen.c                    |  2 +-
 9 files changed, 40 insertions(+), 7 deletions(-)

--- a/hw/xfree86/common/xf86platformBus.c
+++ b/hw/xfree86/common/xf86platformBus.c
@@ -47,6 +47,7 @@
 #include "Pci.h"
 #include "xf86platformBus.h"
 
+#include "randrstr.h"
 int platformSlotClaimed;
 
 int xf86_num_platform_devices;
@@ -470,7 +471,8 @@
    /* attach unbound to 0 protocol screen */
    AttachUnboundGPU(xf86Screens[0]->pScreen, xf86GPUScreens[i]->pScreen);
    xf86AutoConfigOutputDevice(xf86GPUScreens[i], xf86Screens[0]);
-
+   
+   RRTellChanged(xf86Screens[0]->pScreen);
    return 0;
 }
 
@@ -508,12 +510,13 @@
     xf86GPUScreens[i]->pScreen->CloseScreen(xf86GPUScreens[i]->pScreen);
 
     RemoveGPUScreen(xf86GPUScreens[i]->pScreen);
+
     xf86DeleteScreen(xf86GPUScreens[i]);
 
     xf86UnclaimPlatformSlot(&xf86_platform_devices[index], NULL);
 
     xf86_remove_platform_device(index);
-
+    RRTellChanged(xf86Screens[0]->pScreen);
  out:
     return;
 }
--- a/hw/xfree86/modes/xf86Crtc.c
+++ b/hw/xfree86/modes/xf86Crtc.c
@@ -3283,6 +3283,7 @@
         unbound = TRUE;
         AttachOffloadGPU(master->pScreen, pScrn->pScreen);
         slave_config->randr_provider->offload_sink = master_provider;
+	RRSetChanged(master->pScreen);
     }
     if ((master->capabilities & RR_Capability_SourceOutput) &&
                pScrn->capabilities & RR_Capability_SinkOutput) {
@@ -3291,5 +3292,6 @@
             DetachUnboundGPU(pScrn->pScreen);
         AttachOutputGPU(master->pScreen, pScrn->pScreen);
         slave_config->randr_provider->output_source = master_provider;
+	RRSetChanged(master->pScreen);
     }
 }
--- a/hw/xfree86/modes/xf86RandR12.c
+++ b/hw/xfree86/modes/xf86RandR12.c
@@ -1893,15 +1893,20 @@
     xf86CrtcConfigPtr config = XF86_CRTC_CONFIG_PTR(scrn);
     
     if (config->randr_provider == provider) {
+        BOOL detached = FALSE;
         if (config->randr_provider->offload_sink) {
             DetachOffloadGPU(screen);
             config->randr_provider->offload_sink = NULL;
+            RRSetChanged(screen);
+            detached = TRUE;
         }
-        else if (config->randr_provider->output_source) {
+        if (config->randr_provider->output_source) {
             DetachOutputGPU(screen);
             config->randr_provider->output_source = NULL;
+            RRSetChanged(screen);
+            detached = TRUE;
         }
-        else if (screen->current_master)
+        if (!detached && screen->current_master)
             DetachUnboundGPU(screen);
     }
     config->randr_provider = NULL;
--- a/randr/randr.c
+++ b/randr/randr.c
@@ -464,6 +464,28 @@
     return WT_WALKCHILDREN;
 }
 
+void
+RRSetChanged(ScreenPtr pScreen)
+{
+    /* set changed bits on the master screen only */
+    ScreenPtr master;
+    rrScrPriv(pScreen);
+    rrScrPrivPtr mastersp;
+
+    if (pScreen->isGPU) {
+        master = pScreen->current_master;
+	if (!master)
+	  return;
+        mastersp = rrGetScrPriv(master);
+    }
+    else {
+        master = pScreen;
+        mastersp = pScrPriv;
+    }
+
+    mastersp->changed = TRUE;
+}
+
 /*
  * Something changed; send events and adjust pointer position
  */
@@ -484,7 +506,7 @@
         mastersp = pScrPriv;
     }
 
-    if (pScrPriv->changed) {
+    if (mastersp->changed) {
         UpdateCurrentTimeIf();
         if (mastersp->configChanged) {
             mastersp->lastConfigTime = currentTime;
--- a/randr/randrstr.h
+++ b/randr/randrstr.h
@@ -486,6 +486,10 @@
  RRDeliverScreenEvent(ClientPtr client, WindowPtr pWin, ScreenPtr pScreen);
 
 /* randr.c */
+/* set a screen change on the primary screen */
+extern _X_EXPORT void
+RRSetChanged(ScreenPtr pScreen);
+
 /*
  * Send all pending events
  */
--- a/randr/rrcrtc.c
+++ b/randr/rrcrtc.c
@@ -39,7 +39,7 @@
     if (pScreen) {
         rrScrPriv(pScreen);
 
-        pScrPriv->changed = TRUE;
+	RRSetChanged(pScreen);
         /*
          * Send ConfigureNotify on any layout change
          */
--- a/randr/rrinfo.c
+++ b/randr/rrinfo.c
@@ -225,7 +225,7 @@
     pScrPriv->minHeight = minHeight;
     pScrPriv->maxWidth = maxWidth;
     pScrPriv->maxHeight = maxHeight;
-    pScrPriv->changed = TRUE;
+    RRSetChanged(pScreen);
     pScrPriv->configChanged = TRUE;
 }
 
--- a/randr/rroutput.c
+++ b/randr/rroutput.c
@@ -36,7 +36,7 @@
     output->changed = TRUE;
     if (pScreen) {
         rrScrPriv(pScreen);
-        pScrPriv->changed = TRUE;
+	RRSetChanged(pScreen);
         if (configChanged)
             pScrPriv->configChanged = TRUE;
     }
--- a/randr/rrscreen.c
+++ b/randr/rrscreen.c
@@ -143,7 +143,7 @@
     pScrPriv->height = pScreen->height;
     pScrPriv->mmWidth = pScreen->mmWidth;
     pScrPriv->mmHeight = pScreen->mmHeight;
-    pScrPriv->changed = TRUE;
+    RRSetChanged(pScreen);
 /*    pScrPriv->sizeChanged = TRUE; */
 
     RRTellChanged(pScreen);
